api/types/tunnel_strategy.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"encoding/json"
+
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/trace"
+)
+
+const (
+	tunnelStrategyTypeParam     = "type"
+	defaultAgentConnectionCount = 1
+)
+
+// DefaultAgentMeshTunnelStrategy sets default values for a agent mesh
+// tunnel strategy.
+func DefaultAgentMeshTunnelStrategy() *AgentMeshTunnelStrategy {
+	return &AgentMeshTunnelStrategy{}
+}
+
+// DefaultProxyPeeringTunnelStrategy sets default values for a proxy peering
+// tunnel strategy.
+func DefaultProxyPeeringTunnelStrategy() *ProxyPeeringTunnelStrategy {
+	return &ProxyPeeringTunnelStrategy{
+		AgentConnectionCount: defaultAgentConnectionCount,
+	}
+}
+
+// DefaultTunnelStrategy is the default tunnel strategy used when one is not
+// specified.
+func DefaultTunnelStrategy() TunnelStrategy {
+	return &TunnelStrategyV1_AgentMesh{
+		AgentMesh: DefaultAgentMeshTunnelStrategy(),
+	}
+}
+
+// TunnelStrategy defines methods to be implemented by any TunnelStrategy.
+type TunnelStrategy interface {
+	isTunnelStrategyV1_Strategy
+	CheckAndSetDefaults() error
+}
+
+// tunnelStrategyConfig represents a unparsed tunnel strategy configuration.
+type tunnelStrategyConfig struct {
+	Type   TunnelStrategyType     `yaml:"type" json:"type"`
+	Params map[string]interface{} `yaml:",inline" json:"-"`
+}
+
+// newTunnelStrategyConfig creates a new tunnelStrategyConfig instance.
+func newTunnelStrategyConfig() *tunnelStrategyConfig {
+	return &tunnelStrategyConfig{
+		Params: make(map[string]interface{}),
+	}
+}
+
+// setFromMap sets a TunnelStrategyConfig from a map.
+func (c *tunnelStrategyConfig) setFromMap(m map[string]interface{}) error {
+	rawStrategy, ok := m[tunnelStrategyTypeParam]
+	if !ok {
+		return trace.BadParameter("missing type parameter")
+	}
+	strategyType, ok := rawStrategy.(string)
```

Comment:
I don't think it would panic but instead return an error. Either way I will update this to have `getMapCopy` convert the `TunnelStrategyType` back to a string. That way the map representation of the `TunnelStrategyType` is always a string.

Reply:
Fixed here [be0a01b](https://github.com/gravitational/teleport/pull/10223/commits/be0a01bf537e5137f423dd368e9b3c53dafd0ec6)